internal/datanode/compactor.go

```diff
@@ -0,0 +1,549 @@
+	"github.com/milvus-io/milvus/internal/proto/schemapb"
+
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+var (
+	errCompactionTypeUndifined = errors.New("compaction type undefined")
+	errIllegalCompactionPlan   = errors.New("compaction plan illegal")
+	errTransferType            = errors.New("transfer intferface to type wrong")
+	errUnknownDataType         = errors.New("unknown shema DataType")
+)
+
+type iterator = storage.Iterator
+
+type compactor interface {
+	compact() error
+	getPlanID() UniqueID
+}
+
+type compactionTask struct {
+	downloader
+	uploader
+	compactor
+	Replica
+	flushManager
+	allocatorInterface
+
+	dc   types.DataCoord
+	plan *datapb.CompactionPlan
+}
+
+// check if compactionTask implements compactor
+var _ compactor = (*compactionTask)(nil)
+
+func newCompactionTask(
+	dl downloader,
+	ul uploader,
+	replica Replica,
+	fm flushManager,
+	alloc allocatorInterface,
+	dc types.DataCoord,
+	plan *datapb.CompactionPlan) *compactionTask {
+	return &compactionTask{
+		downloader:         dl,
+		uploader:           ul,
+		Replica:            replica,
+		flushManager:       fm,
+		allocatorInterface: alloc,
+		dc:                 dc,
+		plan:               plan,
+	}
+}
+
+func (t *compactionTask) getPlanID() UniqueID {
+	return t.plan.GetPlanID()
+}
+
+func (t *compactionTask) mergeDeltalogs(dBlobs map[UniqueID][]*Blob, timetravelTs Timestamp) (map[UniqueID]Timestamp, *DelDataBuf, error) {
+
+	dCodec := storage.NewDeleteCodec()
+
+	var (
+		pk2ts = make(map[UniqueID]Timestamp)
+		dbuff = &DelDataBuf{
+			delData: &DeleteData{Data: make(map[UniqueID]UniqueID)},
+			tsFrom:  math.MaxUint64,
+			tsTo:    0,
+		}
+	)
+
+	for _, blobs := range dBlobs {
+		_, _, dData, err := dCodec.Deserialize(blobs)
+		if err != nil {
+			log.Warn("merge deltalogs wrong", zap.Error(err))
+			return nil, nil, err
+		}
+
+		for pk, ts := range dData.Data {
+			if Timestamp(ts) >= timetravelTs {
+				pk2ts[pk] = Timestamp(ts)
+				continue
+			}
+
+			dbuff.delData.Data[pk] = ts
+
+			if Timestamp(ts) < dbuff.tsFrom {
+				dbuff.tsFrom = Timestamp(ts)
+			}
+
+			if Timestamp(ts) > dbuff.tsTo {
+				dbuff.tsTo = Timestamp(ts)
+			}
+		}
+	}
+
+	dbuff.updateSize(int64(len(dbuff.delData.Data)))
+
+	return pk2ts, dbuff, nil
+}
+
+func (t *compactionTask) merge(mergeItr iterator, delta map[UniqueID]Timestamp, schema *schemapb.CollectionSchema) (*InsertData, error) {
+
+	var (
+		iData = &InsertData{Data: make(map[storage.FieldID]storage.FieldData)}
+
+		fID2Type    = make(map[UniqueID]schemapb.DataType)
+		fID2Content = make(map[UniqueID][]interface{})
+	)
+
+	for _, fs := range schema.GetFields() {
+		fID2Type[fs.GetFieldID()] = fs.GetDataType()
+	}
+
+	for mergeItr.HasNext() {
+		// There will be no error if HasNext() returns true
+		vInter, _ := mergeItr.Next()
+
+		v, ok := vInter.(*storage.Value)
+		if !ok {
+			log.Warn("transfer interface to Value wrong")
+			return nil, errors.New("Unexpected error")
+		}
+
+		if _, ok := delta[v.ID]; ok {
```

Comment:
we may still have to write the value into file if time travel is not passed.
If user travel back to the old timestamp, even if it is deleted for now,  it may not be deleted by then so  it has to be reserved